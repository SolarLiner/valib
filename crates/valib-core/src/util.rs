//! Utilities for all of `valib`.

use crate::Scalar;
use nalgebra::{
    Dim, Matrix, MatrixView, MatrixViewMut, Storage, StorageMut, Vector, VectorView, VectorViewMut,
    ViewStorage, ViewStorageMut,
};
use num_traits::{AsPrimitive, Float, Zero};
use numeric_literals::replace_float_literals;
use simba::simd::SimdValue;

/// Transmutes a slice into a slice of static arrays, putting the remainder of the slice not fitting
/// as a separate slice.
///
/// # Arguments
///
/// * `data`: Slice to transmute
///
/// returns: (&[[T; N]], &[T])
///
/// # Examples
///
/// ```
/// use valib_core::util::as_nested_arrays;
/// let array = vec![0, 1, 2, 3, 4, 5, 6, 7];
/// let (arrays, remainder) = as_nested_arrays(&array);
/// assert_eq!(&[[0,1,2],[3,4,5]], arrays);
/// assert_eq!(&[6,7], remainder);
/// ```
pub fn as_nested_arrays<T, const N: usize>(data: &[T]) -> (&[[T; N]], &[T]) {
    let rem = data.len() % N;
    let last = data.len() - rem;
    let (complete, remaining) = data.split_at(last);
    let outer_len = complete.len() / N;

    // Safety: Static arrays of N elements have the same representation as N contiguous elements
    let result = unsafe { std::slice::from_raw_parts(complete.as_ptr() as *const _, outer_len) };
    (result, remaining)
}

/// Transmutes a mutable slice into a slice of static arrays, putting the remainder of the slice not
/// fitting as a separate slice.
///
/// # Arguments
///
/// * `data`: Slice to transmute
///
/// returns: (&mut [[T; N]], &mut [T])
///
/// # Examples
///
/// ```
/// use valib_core::util::as_nested_arrays_mut;
/// let mut array = vec![0, 1, 2, 3, 4, 5, 6, 7];
/// let (arrays, remainder) = as_nested_arrays_mut(&mut array);
/// assert_eq!(&[[0,1,2],[3,4,5]], arrays);
/// assert_eq!(&[6,7], remainder);
///
/// arrays[1][0] = 10;
/// assert_eq!(vec![0, 1, 2, 10, 4, 5, 6, 7], array);
/// ```
pub fn as_nested_arrays_mut<T, const N: usize>(data: &mut [T]) -> (&mut [[T; N]], &mut [T]) {
    let rem = data.len() % N;
    let last = data.len() - rem;
    let (complete, remaining) = data.split_at_mut(last);
    let outer_len = complete.len() / N;

    // Safety: Static arrays of N elements have the same representation as N contiguous elements
    let result =
        unsafe { std::slice::from_raw_parts_mut(complete.as_mut_ptr() as *mut _, outer_len) };
    (result, remaining)
}

/// Index a slice of scalars with a SIMD index, returning a SIMD with the corresponding scalar values
/// on each lane.
///
/// # Arguments
///
/// * `values`: Slice to index
/// * `index`: Index SIMD value
///
/// returns: Simd
///
/// # Examples
///
/// ```
/// use simba::simd::{AutoF32x4, AutoUsizex4};
/// use valib_core::util::simd_index_scalar;
/// let data = [0.0, 0.1, 0.2, 0.3];
/// let index = AutoUsizex4::new(0, 1, 2, 3);
/// let  ret = simd_index_scalar::<AutoF32x4, _>(&data, index);
/// assert_eq!(AutoF32x4::new(0.0, 0.1, 0.2, 0.3), ret);
/// ```
pub fn simd_index_scalar<Simd: Zero + SimdValue, Index: SimdValue<Element = usize>>(
    values: &[Simd::Element],
    index: Index,
) -> Simd
where
    Simd::Element: Copy,
{
    let mut ret = Simd::zero();
    for i in 0..Simd::LANES {
        let ix = index.extract(i);
        ret.replace(i, values[ix]);
    }
    ret
}

/// Index a slice of SIMD values with a SIMD index, returning a new SIMD values corresponding to
/// each lane of the values for each index.
///
/// # Arguments
///
/// * `values`: Slice to index
/// * `index`: Index SIMD value
///
/// returns: Simd
///
/// # Examples
///
/// ```
/// use simba::simd::{AutoF32x2, AutoUsizex2};
/// use valib_core::util::simd_index_simd;
/// let data = [AutoF32x2::new(0.0, 0.1), AutoF32x2::new(1.0, 1.1)];
/// let index = AutoUsizex2::new(1, 0);
/// let ret = simd_index_simd(&data, index);
/// assert_eq!(AutoF32x2::new(1.0, 0.1), ret);
/// ```
pub fn simd_index_simd<Simd: Zero + SimdValue, Index: SimdValue>(
    values: &[Simd],
    index: Index,
) -> Simd
where
    <Index as SimdValue>::Element: AsPrimitive<usize>,
{
    let mut ret = Simd::zero();
    for i in 0..Index::LANES {
        let ix = index.extract(i).as_();
        ret.replace(i, values[ix].extract(i));
    }
    ret
}

/// Generic method for checking if a SIMD float is finie or not. Returns a SIMD mask where the lanes
/// are finite.
///
/// # Arguments
///
/// * `value`: Value to check
///
/// returns: <Simd as SimdValue>::SimdBool
///
/// # Examples
///
/// ```
/// use simba::simd::{AutoF32x4, AutoBoolx4};
/// use valib_core::util::simd_is_finite;
/// let value = AutoF32x4::new(0.0, f32::INFINITY, f32::NAN, f32::EPSILON);
/// let ret = simd_is_finite(value);
/// assert_eq!(AutoBoolx4::new(true, false, false, true), ret);
/// ```
pub fn simd_is_finite<
    Simd: SimdValue<Element: Float, SimdBool: Default + SimdValue<Element = bool>>,
>(
    value: Simd,
) -> Simd::SimdBool {
    let mut mask = Simd::SimdBool::default();
    for i in 0..Simd::LANES {
        mask.replace(i, value.extract(i).is_finite())
    }
    mask
}

/// Shortcut function to perform linear interpolation. Uses the interpolation module.
pub fn lerp<T: Scalar>(t: T, a: T, b: T) -> T {
    use crate::math::interpolation::{Interpolate, Linear};
    Linear.interpolate(t, [a, b])
}

/// Computes the frequency of a MIDI note number, assuming 12TET and A4 = 440 Hz
///
/// # Arguments
///
/// * `midi_note`: MIDI note number
///
/// returns: T
#[replace_float_literals(T::from_f64(literal))]
pub fn midi_to_freq<T: Scalar>(midi_note: u8) -> T {
    440.0 * semitone_to_ratio(T::from_f64(midi_note as _) - 69.0)
}

/// Compute the ratio corresponding to the given semitone change, such that multiplying a frequency
/// by this value changes it by the given semitones.
///
/// # Arguments
///
/// * `semi`: Semitone change
///
/// returns: T
#[replace_float_literals(T::from_f64(literal))]
pub fn semitone_to_ratio<T: Scalar>(semi: T) -> T {
    2.0.simd_powf(semi / 12.0)
}

pub fn matrix_view<T: Scalar, R: Dim, C: Dim, S: Storage<T, R, C>>(
    m: &Matrix<T, R, C, S>,
) -> MatrixView<T, R, C, S::RStride, S::CStride> {
    MatrixView::from_data(unsafe {
        let shape = m.shape_generic();
        ViewStorage::new_unchecked(&m.data, (0, 0), shape)
    })
}

pub fn matrix_view_mut<T: Scalar, R: Dim, C: Dim, S: StorageMut<T, R, C>>(
    m: &mut Matrix<T, R, C, S>,
) -> MatrixViewMut<T, R, C, S::RStride, S::CStride> {
    MatrixViewMut::from_data(unsafe {
        let shape = m.shape_generic();
        ViewStorageMut::new_unchecked(&mut m.data, (0, 0), shape)
    })
}

pub fn vector_view<T: Scalar, D: Dim, S: Storage<T, D>>(
    v: &Vector<T, D, S>,
) -> VectorView<T, D, S::RStride, S::CStride> {
    VectorView::from_data(unsafe {
        let shape = v.shape_generic();
        ViewStorage::new_unchecked(&v.data, (0, 0), shape)
    })
}

pub fn vector_view_mut<T: Scalar, D: Dim, S: StorageMut<T, D>>(
    v: &mut Vector<T, D, S>,
) -> VectorViewMut<T, D, S::RStride, S::CStride> {
    VectorViewMut::from_data(unsafe {
        let shape = v.shape_generic();
        ViewStorageMut::new_unchecked(&mut v.data, (0, 0), shape)
    })
}

#[cfg(feature = "test-utils")]
pub mod tests;
