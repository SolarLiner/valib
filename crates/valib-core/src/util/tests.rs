//! Test utilities. Needs the `test-utils` feature to enable this module.
use std::{ops::Range, path::Path};

use plotters::coord::{self, ranged1d::ValueFormatter};
use plotters::{chart::SeriesAnno, prelude::*};

fn assert_ok(res: Result<(), impl std::fmt::Display>) {
    match res {
        Ok(()) => {}
        Err(value) => panic!("Not OK: {value}"),
    }
}

/// Single time/frequency series
pub struct Series<'a> {
    /// Label of the series
    pub label: &'a str,
    /// Sample rate of the time series
    pub samplerate: f32,
    /// Y-values of the series
    pub series: &'a [f32],
    /// Display color,
    pub color: &'a RGBColor,
}

impl<'a> Series<'a> {
    /// Validate that the series is well-formed
    pub fn validate(&self) -> Result<(), String> {
        if self.samplerate <= 0. {
            return Err(format!("Series: {:?}: Samplerate is negative", self.label));
        }
        if self.series.is_empty() {
            return Err(format!("Series: {:?}: No data", self.label));
        }

        Ok(())
    }

    /// Compute the range of x values spanning this series.
    ///
    /// # Arguments
    ///
    /// * `bode`: The series represent frequency data as opposed to time data.
    ///
    /// returns: Range<f32>
    pub fn timescale(&self, bode: bool) -> Range<f32> {
        assert_ok(self.validate());
        if bode {
            0.0..self.samplerate / 2.0
        } else {
            let tmax = self.series.len() as f32 / self.samplerate;
            0.0..tmax
        }
    }

    /// Return the y-axis range that this series span
    pub fn y_range(&self) -> Range<f32> {
        assert_ok(self.validate());
        let min = self.series.iter().copied().min_by(f32::total_cmp).unwrap();
        let max = self.series.iter().copied().max_by(f32::total_cmp).unwrap();
        min..max
    }

    fn as_series<DB: DrawingBackend>(&self, bode: bool) -> LineSeries<DB, (f32, f32)> {
        LineSeries::new(
            self.series.iter().copied().enumerate().map(|(i, y)| {
                let x = if bode {
                    i as f32
                } else {
                    i as f32 / self.samplerate
                };
                (x, y)
            }),
            self.color,
        )
    }

    fn apply_legend(&self, ann: &mut SeriesAnno<impl DrawingBackend>) {
        let color = *self.color;
        ann.label(self.label);
        ann.legend(move |(x, y)| PathElement::new([(x, y), (x + 20, y)], color));
    }
}

/// Simple, high-level abstraction to plotting time/frequency series
pub struct Plot<'a> {
    /// Title of the graph
    pub title: &'a str,
    /// Is the graph a Bode plot (plotting frequencies) or a time plot ?
    pub bode: bool,
    /// List of series to plot
    pub series: &'a [Series<'a>],
}

impl<'a> Plot<'a> {
    /// Validate that the plot is well-formed.
    pub fn validate(&self) -> Result<(), String> {
        if self.series.is_empty() {
            return Err(format!("Plot {:?}: no series", self.title));
        }
        self.series.iter().try_for_each(|s| s.validate())?;
        Ok(())
    }

    /// Render the plot into the provided drawing area. This is a lower-level method that allows you
    /// to direct rendering wherever you want.
    ///
    /// # Arguments
    ///
    /// * `output`: Output drawing area, on which the plot will be drawn.
    ///
    /// returns: ()
    pub fn render_into(&self, output: &DrawingArea<impl DrawingBackend, coord::Shift>) {
        use plotters::prelude::*;
        assert_ok(self.validate());

        let timescale = self
            .series
            .iter()
            .map(|s| s.timescale(self.bode))
            .reduce(|l, r| {
                let start = l.start.min(r.start);
                let end = l.end.max(r.end);
                start..end
            })
            .unwrap();
        let timescale = if self.bode {
            timescale.start * 2.0..timescale.end * 2.0
        } else {
            timescale
        };

        let yrange = self
            .series
            .iter()
            .map(|s| s.y_range())
            .reduce(|l, r| {
                let start = l.start.min(r.start);
                let end = l.end.max(r.end);
                start..end
            })
            .unwrap();

        let mut ctx = ChartBuilder::on(output);
        ctx.set_label_area_size(LabelAreaPosition::Left, 40)
            .set_label_area_size(LabelAreaPosition::Bottom, 40)
            .caption(self.title, ("sans-serif", 40));
        if self.bode {
            let ctx = ctx
                .build_cartesian_2d(timescale.log_scale(), yrange.log_scale())
                .unwrap();
            self.render(ctx);
        } else {
            let ctx = ctx.build_cartesian_2d(timescale, yrange).unwrap();
            self.render(ctx);
        }
    }

    fn render<'ctx, T: 'ctx + Ranged<ValueType = f32> + ValueFormatter<f32>>(
        &self,
        mut ctx: ChartContext<'ctx, impl 'ctx + DrawingBackend, Cartesian2d<T, T>>,
    ) {
        ctx.configure_mesh().draw().unwrap();

        for series in self.series {
            let ann = ctx.draw_series(series.as_series(self.bode)).unwrap();
            series.apply_legend(ann);
        }

        ctx.configure_series_labels()
            .background_style(WHITE.mix(0.8))
            .draw()
            .unwrap();
    }

    /// Create an SVG file in which this plot is going to be rendered.
    ///
    /// # Arguments
    ///
    /// * `filename`: Filename pointing to the generated SVG file
    ///
    /// returns: ()
    pub fn create_svg(&self, filename: impl AsRef<Path>) {
        let path = filename.as_ref();
        let _ = std::fs::create_dir_all(path.parent().expect("Filename is empty"));
        let root = SVGBackend::new(path, (600, 400)).into_drawing_area();
        root.fill(&WHITE).unwrap();
        self.render_into(&root);
    }
}
