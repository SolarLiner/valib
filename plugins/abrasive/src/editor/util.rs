use directories::ProjectDirs;
use nih_plug::nih_log;
use std::env;
use std::path::{Path, PathBuf};

pub fn resolve_asset_file(file: &Path) -> PathBuf {
    search_dirs()
        .map(|p| p.join(file))
        .inspect(|p| nih_log!("Getting asset: trying {}", p.display()))
        .find(|p| p.is_file())
        .expect(&format!("Cannot find assets directory {}", file.display()))
}

pub fn resolve_asset_dir(dir: &Path) -> PathBuf {
    search_dirs()
        .map(|p| p.join(dir))
        .inspect(|p| nih_log!("Getting assets directory: trying {}", p.display()))
        .find(|p| p.is_dir())
        .expect(&format!("Cannot find assets directory {}", dir.display()))
}

fn search_dirs() -> impl Iterator<Item = PathBuf> {
    [
        installed_assets_dir(),
        dev_workspace_dir(),
        plugin_package_workspace_dir(),
        current_dir(),
        current_exe(),
    ]
    .into_iter()
    .flatten()
}

fn installed_assets_dir() -> Option<PathBuf> {
    let dirs = ProjectDirs::from("dev.solarliner", "valib", env!("CARGO_PKG_NAME"))?;
    Some(dirs.data_dir().join("assets"))
}

fn dev_workspace() -> Option<PathBuf> {
    if cfg!(feature = "distribution") {
        None
    } else {
        // Plugin project is at plugins/abrasive, so we need to climb up twice to get the workspace
        Some(
            Path::new(env!("CARGO_MANIFEST_DIR"))
                .parent()?
                .parent()?
                .to_path_buf(),
        )
    }
}

fn dev_workspace_dir() -> Option<PathBuf> {
    dev_workspace().map(|p| p.join("assets"))
}

fn plugin_package_workspace_dir() -> Option<PathBuf> {
    dev_workspace().map(|dir| {
        dir.join("plugins")
            .join(env!("CARGO_PKG_NAME"))
            .join("assets")
    })
}

fn current_dir() -> Option<PathBuf> {
    env::current_dir().ok().map(|p| p.join("assets"))
}

fn current_exe() -> Option<PathBuf> {
    env::current_exe()
        .ok()
        .and_then(|p| Some(p.canonicalize().ok()?.parent()?.join("assets")))
}
