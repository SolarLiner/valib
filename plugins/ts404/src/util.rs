use std::collections::VecDeque;
use nih_plug::nih_log;
use num_traits::Zero;
use valib::Scalar;

#[derive(Debug, Clone)]
pub struct Rms<T> {
    data: VecDeque<T>,
    summed_squared: T,
}

impl<T: Zero> Rms<T> {
    pub fn new(size: usize) -> Self {
        Self {
            data: (0..size).map(|_| T::zero()).collect(),
            summed_squared: T::zero(),
        }
    }
}

impl<T: Scalar> Rms<T> {
    pub fn add_element(&mut self, value: T) -> T {
        let v2 = value.simd_powi(2);
        self.summed_squared -= self.data.pop_front().unwrap();
        self.summed_squared += v2;
        self.data.push_back(v2);
        self.get_rms()
    }
    
    pub fn get_rms(&self) -> T {
        self.summed_squared.simd_sqrt()
    }
}