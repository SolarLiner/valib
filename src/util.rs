use num_traits::{FromPrimitive, Num, One, Zero, AsPrimitive};
use numeric_literals::replace_float_literals;
use simba::simd::{SimdPartialOrd, SimdValue};

use crate::{Scalar, SimdCast};

pub fn simd_index_scalar<Simd: Zero + SimdValue, Index: SimdValue<Element = usize>>(
    values: &[Simd::Element],
    index: Index,
) -> Simd
where
    Simd::Element: Copy,
{
    let mut ret = Simd::zero();
    for i in 0..Simd::lanes() {
        let ix = index.extract(i);
        ret.replace(i, values[ix]);
    }
    ret
}

pub fn simd_index_simd<Simd: Zero + SimdValue, Index: SimdValue>(
    values: &[Simd],
    index: Index,
) -> Simd where <Index as SimdValue>::Element: AsPrimitive<usize> {
    let mut ret = Simd::zero();
    for i in 0..Index::lanes() {
        let ix = index.extract(i).as_();
        ret.replace(i, values[ix].extract(i));
    }
    ret
}

#[replace_float_literals(T::from_f64(literal))]
pub fn lerp_block<T: Scalar + SimdCast<usize>>(out: &mut [T], inp: &[T])
where
    <T as SimdCast<usize>>::Output: Copy + Num + FromPrimitive + SimdPartialOrd,
{
    let ix_max = <T as SimdCast<usize>>::Output::from_usize(inp.len() - 1).unwrap();
    let rate = T::from_f64(inp.len() as f64) / T::from_f64(out.len() as f64);

    for (i, y) in out.iter_mut().enumerate() {
        let j = T::from_f64(i as f64) * rate;
        let f = j.simd_fract();
        let j = j.simd_floor().cast();
        let jp1 = (j + <T as SimdCast<usize>>::Output::one()).simd_min(ix_max);
        let a = simd_index_simd(inp, j);
        let b = simd_index_simd(inp, jp1);
        *y = lerp(f, a, b);
    }
}

pub fn lerp<T: Scalar>(t: T, a: T, b: T) -> T {
    a + t * (b - a)
}

#[cfg(test)]
mod tests {
    use super::lerp_block;

    #[test]
    fn interp_block() {
        let a = [0., 1., 1.];
        let mut actual = [0.; 12];
        let expected = [0., 0.25, 0.5, 0.75, 1., 1., 1., 1., 1., 1., 1., 1.];
        lerp_block(&mut actual, &a);
        assert_eq!(actual, expected);
    }
}
