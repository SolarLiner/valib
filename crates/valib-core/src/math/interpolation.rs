use nalgebra::SimdPartialOrd;
use num_traits::{FromPrimitive, NumAssignOps, NumOps};
use numeric_literals::replace_float_literals;
use simba::simd::SimdValue;

use crate::util::simd_index_simd;
use crate::{Scalar, SimdCast};

pub trait SimdIndex:
    Copy + NumAssignOps + NumOps + SimdPartialOrd + SimdValue<Element = usize>
{
}

impl<T: Copy + NumAssignOps + NumOps + SimdPartialOrd + SimdValue<Element = usize>> SimdIndex
    for T
{
}

pub trait SimdInterpolatable: SimdCast<usize>
where
    <Self as SimdCast<usize>>::Output: SimdIndex,
{
    fn index_from_usize(value: usize) -> Self::Output {
        Self::Output::splat(value)
    }
}

impl<T: SimdCast<usize>> SimdInterpolatable for T where <T as SimdCast<usize>>::Output: SimdIndex {}

/// Interpolation trait. Interpolators implement function that connect discrete points into a continuous function.
/// Functions can have any number of taps, both in the forward and bacward directions. It's up to the called to provide
/// either actual existing points or extrapolated values when the indices would be out of bounds.
pub trait Interpolate<T, const N: usize> {
    /// Provide the indices needed for interpolating around the input index, where the index is the element that corresponds to t = 0.
    fn indices(index: usize) -> [usize; N];

    /// Interpolate a single point from the given taps (in the same order as the indices defined in [`Interpolate::indices`]).
    /// The t parameter is assumed to be in the 0..=1 range, and it's up to the caller to provide values in the valid range.
    fn interpolate(&self, t: T, taps: [T; N]) -> T;

    /// Interpolate a value from an entire slice, where the t parameter is the "floating index" into the slice
    /// (meaning 3.5 is halfway between index 3 and 4 on the given slice).
    #[profiling::function]
    fn interpolate_on_slice(&self, t: T, values: &[T]) -> T
    where
        T: Scalar + SimdInterpolatable,
        <T as SimdCast<usize>>::Output: SimdIndex,
    {
        let input_frac = t.simd_fract();
        let input_index = t.simd_floor().cast();
        let taps_ix: [_; N] = std::array::from_fn(|i| {
            let mut output = input_index;
            for j in 0..<T as SimdCast<usize>>::Output::LANES {
                output.replace(j, Self::indices(output.extract(j))[i]);
            }
            output
        });

        let zero = T::index_from_usize(0);
        let ix_max = T::index_from_usize(values.len() - 1);
        let taps = taps_ix.map(|tap| simd_index_simd(values, tap.simd_clamp(zero, ix_max)));
        self.interpolate(input_frac, taps)
    }

    /// Interpolate one slice into another, where the output slice ends up containing the same "range" of values as
    /// the input slice, but also automatically performs interpolation using this instance.
    #[profiling::function]
    fn interpolate_slice(&self, output: &mut [T], input: &[T])
    where
        T: Scalar + SimdInterpolatable,
        <T as SimdCast<usize>>::Output: SimdIndex,
    {
        let rate = input.len() as f64 / output.len() as f64;

        for (i, o) in output.iter_mut().enumerate() {
            let t = T::from_f64(rate * i as f64);
            *o = self.interpolate_on_slice(t, input);
        }
    }
}

/// Zero-hold interpolation, where the output is the input without additional computation.
#[derive(Debug, Copy, Clone)]
pub struct ZeroHold;

impl<T> Interpolate<T, 1> for ZeroHold {
    fn indices(index: usize) -> [usize; 1] {
        [index]
    }

    fn interpolate(&self, _: T, [x]: [T; 1]) -> T {
        x
    }
}

/// Nearest-neighbor interpolation, where the neighbor is chosen based on how close it is to the floating index.
#[derive(Debug, Copy, Clone)]
pub struct Nearest;

impl<T: SimdPartialOrd + FromPrimitive> Interpolate<T, 2> for Nearest {
    fn indices(index: usize) -> [usize; 2] {
        [index, index + 1]
    }

    fn interpolate(&self, t: T, [a, b]: [T; 2]) -> T {
        b.select(t.simd_gt(T::from_f64(0.5).unwrap()), a)
    }
}

/// Standard-issue linear interpolation algorithm.
#[derive(Debug, Copy, Clone)]
pub struct Linear;

impl<T: Scalar> Interpolate<T, 2> for Linear {
    fn indices(index: usize) -> [usize; 2] {
        [index, index + 1]
    }

    fn interpolate(&self, t: T, [a, b]: [T; 2]) -> T {
        a + (b - a) * t
    }
}

/// Sine interpolation, which is linear interpolation where the control input is first modulated by
/// a cosine function. Produes smoother results than bare linear interpolation because the interpolation
/// is smooth at the limits, as all derivatives are 0 there.
#[derive(Debug, Copy, Clone)]
pub struct MappedLinear<F>(pub F);

pub fn sine_interpolation<T: Scalar>() -> MappedLinear<impl Fn(T) -> T> {
    MappedLinear(|t| T::simd_cos(t * T::simd_pi()))
}

impl<T, F: Fn(T) -> T> Interpolate<T, 2> for MappedLinear<F>
where
    Linear: Interpolate<T, 2>,
{
    fn indices(index: usize) -> [usize; 2] {
        Linear::indices(index)
    }

    fn interpolate(&self, t: T, taps: [T; 2]) -> T {
        Linear.interpolate(self.0(t), taps)
    }
}

/// Cubic algorithm, smoother than [`Linear`], but needs 4 taps instead of 2.
#[derive(Debug, Copy, Clone)]
pub struct Cubic;

impl<T: Scalar> Interpolate<T, 4> for Cubic {
    fn indices(index: usize) -> [usize; 4] {
        [index.saturating_sub(1), index, index + 1, index + 2]
    }

    #[replace_float_literals(T::from_f64(literal))]
    fn interpolate(&self, t: T, taps: [T; 4]) -> T {
        taps[1]
            + 0.5
                * t
                * (taps[2] - taps[0]
                    + t * (2.0 * taps[0] - 5.0 * taps[1] + 4.0 * taps[2] - taps[3]
                        + t * (3.0 * (taps[1] - taps[2]) + taps[3] - taps[0])))
    }
}

/// 4-tap cubic Hermite spline interpolation
#[derive(Debug, Copy, Clone)]
pub struct Hermite;

impl<T: Scalar> Interpolate<T, 4> for Hermite {
    fn indices(index: usize) -> [usize; 4] {
        [index.saturating_sub(1), index, index + 1, index + 2]
    }

    #[replace_float_literals(T::from_f64(literal))]
    fn interpolate(&self, t: T, taps: [T; 4]) -> T {
        let c0 = taps[1];
        let c1 = 0.5 * (taps[2] - taps[0]);
        let c2 = taps[0] - 2.5 * taps[1] + 2.0 * taps[2] - 0.5 * taps[3];
        let c3 = 1.5 * (taps[1] - taps[2]) + 0.5 * (taps[3] - taps[0]);
        ((c3 * t + c2) * t + c1) * t + c0
    }
}

#[derive(Debug, Copy, Clone)]
pub struct Lanczos;

impl<T: Scalar> Interpolate<T, 7> for Lanczos {
    fn indices(index: usize) -> [usize; 7] {
        std::array::from_fn(|i| (index + i).saturating_sub(3))
    }

    fn interpolate(&self, t: T, taps: [T; 7]) -> T {
        let a = T::from_f64(3.0);
        let x = T::from_f64(4.0) + t;
        taps.iter()
            .copied()
            .enumerate()
            .map(|(i, s)| {
                let i = x.simd_floor() - a + T::one() + T::from_f64(i as _);
                s * Self::window::<T>(x - i)
            })
            .reduce(|a, b| a + b)
            .unwrap_or_else(T::zero)
    }
}

impl Lanczos {
    fn window<T: Scalar>(t: T) -> T {
        fn sinc<T: Scalar>(x: T) -> T {
            let nx = x * T::simd_pi();
            let y = nx.simd_sin() / nx;
            T::one().select(x.simd_eq(T::zero()), y)
        }

        sinc(t) * sinc(t / T::from_f64(3.0))
    }
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use super::*;

    #[rstest]
    fn test_interpolate<Interp, const TAPS: usize>(
        #[values(
            ZeroHold,
            Nearest,
            Linear,
            Cubic,
            Hermite,
            sine_interpolation(),
            Lanczos
        )]
        interp: Interp,
    ) where
        Interp: Interpolate<f64, TAPS>,
    {
        let a = [0., 1., 1.];
        let mut actual = [0.; 12];
        interp.interpolate_slice(&mut actual, &a);
        let name = format!(
            "test_interpolate_{}",
            std::any::type_name::<Interp>()
                .replace(['<', '>'], "__")
                .replace("::", "__")
        );
        insta::assert_csv_snapshot!(name, &actual as &[_], { "[]" => insta::rounded_redaction(6) });
    }
}
