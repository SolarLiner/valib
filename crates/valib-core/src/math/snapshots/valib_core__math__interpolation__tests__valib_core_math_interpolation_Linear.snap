---
source: crates/valib-core/src/math/interpolation.rs
expression: "&actual as &[_]"
---
0.0
0.25
0.5
0.75
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
